scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/ParquetBucketMetadata.java

```diff
@@ -0,0 +1,266 @@
+        && Arrays.equals(keyPath, that.keyPath);
+  }
+
+  @Override
+  public K extractKey(V value) {
+    if (recordType == null) {
+      recordType = getRecordType(value.getClass());
+    }
+    switch (recordType) {
+      case AVRO:
+        return extractAvroKey(value);
+      case SCALA:
+        return extractScalaKey(value);
+    }
+    throw new IllegalStateException("Should never reach here");
+  }
+
+  private K extractAvroKey(V value) {
+    GenericRecord node = (GenericRecord) value;
+    for (int i = 0; i < keyPath.length - 1; i++) {
+      node = (GenericRecord) node.get(keyPath[i]);
+    }
+    @SuppressWarnings("unchecked")
+    K key = (K) node.get(keyPath[keyPath.length - 1]);
+    return key;
+  }
+
+  // FIXME: what about `Option[T]`
+  private K extractScalaKey(V value) {
+    if (getters == null) {
+      getters = new Method[keyPath.length];
+      Class<?> cls = value.getClass();
+      for (int i = 0; i < keyPath.length; i++) {
+        Method getter = null;
+        try {
+          getter = cls.getMethod(keyPath[i]);
+        } catch (NoSuchMethodException e) {
+          throw new IllegalStateException(
+              String.format("Failed to prepare getter %s for class %s", keyPath[i], cls));
+        }
+        getters[i] = getter;
+        cls = getter.getReturnType();
+      }
+    }
+    Object obj = value;
+    for (Method getter : getters) {
+      try {
+        obj = getter.invoke(obj);
+      } catch (IllegalAccessException | InvocationTargetException e) {
+        throw new IllegalStateException(
+            String.format("Failed to get field %s from class %s", getter.getName(), obj));
+      }
+    }
+    @SuppressWarnings("unchecked")
+    K key = (K) obj;
+    return key;
+  }
+
+  private static String[] toKeyPath(String keyField) {
+    return keyField.split("\\.");
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Logic for dealing with Avro records vs Scala case classes
+  ////////////////////////////////////////////////////////////////////////////////
+
+  private static RecordType getRecordType(Class<?> recordClass) {
+    if (GenericRecord.class.isAssignableFrom(recordClass)) {
+      return RecordType.AVRO;
+    } else if (scala.Product.class.isAssignableFrom(recordClass)) {
+      return RecordType.SCALA;
+    } else {
+      throw new IllegalArgumentException("Unsupported record class " + recordClass.getName());
+    }
+  }
+
+  private static String validateKeyField(String keyField, Class<?> keyClass, Class<?> recordClass) {
+    switch (getRecordType(recordClass)) {
+      case AVRO:
+        return AvroUtils.validateKeyField(
+            keyField,
+            keyClass,
+            new ReflectData(recordClass.getClassLoader()).getSchema(recordClass));
+      case SCALA:
+        return validateScalaKeyField(keyField, keyClass, recordClass);
+    }
+    throw new IllegalStateException("Should never reach here");
```

Comment:
shouldn't this go in the `default` case of the switch statement?

Reply:
Brain fart, Pavlovian response IntelliJ squiggly line at the bottom 🐕